src/components/smart_objects/src/object_schema_item.cc

```diff
@@ -157,6 +158,39 @@ errors::eType CObjectSchemaItem::validate(
   return errors::OK;
 }
 
+bool CObjectSchemaItem::filterInvalidEnums(
```

Comment:
Consider following approach. 
```cpp


bool CObjectSchemaItem::filterInvalidEnums(
    SmartObject& Object,
    const utils::SemanticVersion& MessageVersion,
    rpc::ValidationReport* report) {
  
  auto filter_invalid_enums = [this, &MessageVersion, report](member) {
      const SMember* member =
          GetCorrectMember(members_it->second, MessageVersion);
      rreturn member->mSchemaItem->filterInvalidEnums(
              Object[key], MessageVersion, &report->ReportSubobject(key));
  };

  auto should_be_erazed = [](member) -> bool {
        // The member is safe to filter if it is non-mandatory
        // Object is no longer valid if the member is mandatory.
        // To maintain the general structure of the message, only leaf nodes
        // (individual enum values) should be filtered in this case.
      return member->mSchemaItem->GetType() == TYPE_ENUM && !member->mIsMandatory;
  };

  bool valid = true;
  for (const auto& key : Object.enumerate()) {
    auto members_it = mMembers.find(key);
    DCHECK_OR_RETURN(mMembers.end() != members_it, false);  //  object should match schema
    if (filter_invalid_enums(*members_it)) {
        if (member->mIsMandatory) {
          valid = false;
        }
        if (should_be_erazed(member)) {
            Object.erase(key); 
        }
    }
  }
  return !valid;
}
```

The same as above, I didn't check it and it not so performant. but I feel it as more readable.

Reply:
Seems a lambda would be overkill for "should_be_erazed" when a local variable would work fine. IMO unnecessary lambdas make things less clear overall. 

I'll see what I can do to clean it up.

Reply:
Refactored in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/b4452abf1c2f607f6dc68cb96889cdaff8363dde